ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/FreshnessFilter.java

```diff
@@ -0,0 +1,104 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.execution.streams.RoutingFilter;
+import io.confluent.ksql.execution.streams.RoutingOptions;
+import io.confluent.ksql.rest.entity.LagInfoEntity;
+import io.confluent.ksql.rest.entity.QueryStateStoreId;
+import io.confluent.ksql.util.KsqlHost;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.function.Function;
+import org.apache.kafka.streams.state.HostInfo;
+
+/**
+ * A RoutingFilter that filters hosts based upon changelog processing lag.
+ */
+public class FreshnessFilter implements RoutingFilter {
+
+  private final ImmutableMap<KsqlHost, Optional<LagInfoEntity>> lagByHost;
+  private final RoutingOptions routingOptions;
+  private final OptionalLong maxEndOffset;
+
+  private FreshnessFilter(
+      final RoutingOptions routingOptions,
+      final ImmutableMap<KsqlHost, Optional<LagInfoEntity>> lagByHost,
+      final OptionalLong maxEndOffset
+  ) {
+    this.lagByHost = requireNonNull(lagByHost, "lagByHost");
+    this.routingOptions = requireNonNull(routingOptions, "requireNonNull");
+    this.maxEndOffset = requireNonNull(maxEndOffset, "maxEndOffset");
+  }
+
+  @Override
+  public boolean filter(final HostInfo activeHost, final KsqlHost hostInfo) {
+    final long offsetLagAllowed = routingOptions.getOffsetLagAllowed();
+    if (offsetLagAllowed >= 0) {
+      return lagByHost.getOrDefault(hostInfo, Optional.empty())
+          .map(hostLag -> {
+            // Compute the lag from the maximum end offset we've seen
+            long endOffset = maxEndOffset.orElse(hostLag.getEndOffsetPosition());
+            long offsetLag = Math.max(endOffset - hostLag.getCurrentOffsetPosition(), 0);
+            return offsetLag <= offsetLagAllowed;
+          })
+          // If we don't have lag info, we'll be conservative and include the host
+          .orElse(true);
+    }
+    return true;
+  }
+
+  /**
+   * Creates a FreshnessFilter
+   * @param lagReportingAgent The optional lag reporting agent.
+   * @param routingOptions The routing options
+   * @param hosts The set of all hosts that have the store, including actives and standbys
+   * @param applicationQueryId The application query id
+   * @param storeName The state store name
+   * @param partition The partition of the topic
+   * @return a new FreshnessFilter, unless lag reporting is disabled.
+   */
+  public static Optional<FreshnessFilter> create(
+      final Optional<LagReportingAgent> lagReportingAgent,
+      final RoutingOptions routingOptions,
+      final List<KsqlHost> hosts,
```

Comment:
IIRC.. we have a stream already in KsLocator and we make it a list and pass it in? we seem to be recreating a stream out of it anyway.. So change hosts to `Stream<KsqlHost>`?

Reply:
You can only process a stream once, so I have to stick with a list here.  I removed the stream in KsLocator anyway and just produce a single list.